src/reference_implementations/fairness_measurement/crow_s_pairs/llm_crows_pairs.ipynb

```diff
@@ -447,77 +441,200 @@
+     "output_type": "stream",
+     "text": [
+      "Something went wrong in activation retrieval...retrying\n"
+     ]
+    },
+    {
+     "name": "stderr",
+     "output_type": "stream",
+     "text": [
+      " 93%|█████████▎| 1400/1508 [49:46<04:12,  2.33s/it]"
+     ]
+    },
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Something went wrong in activation retrieval...retrying\n"
+     ]
+    },
+    {
+     "name": "stderr",
+     "output_type": "stream",
+     "text": [
+      "100%|██████████| 1508/1508 [53:31<00:00,  2.13s/it]\n"
      ]
     }
    ],
    "source": [
     "neutral = 0\n",
     "\n",
-    "# The score for a sentence is the sum of log probability of each word in the sentence.\n",
-    "\n",
     "total = len(df_data.index)\n",
+    "# Determines the threshold for a \"neutral\" estimation\n",
+    "epsilon = 0.01\n",
     "with tqdm(total=total) as pbar:\n",
     "    for index, data in df_data.iterrows():\n",
     "        direction = data[\"direction\"]\n",
     "        bias = data[\"bias_type\"]\n",
-    "        likelihoods = []\n",
     "\n",
-    "        prompt_sentences = [data[\"advantage_group\"], data[\"disadvantage_group\"]]\n",
-    "        # Function to focus on extracting log probabilities and tokens based on the sentences.\n",
-    "        activations = model.get_activations(prompt_sentences, [], generation_config)\n",
-    "        for logprobs, tokens in zip(activations.logprobs, activations.tokens):\n",
-    "            # indexing \"</s>\" token from the list of tokens given from the model output\n",
-    "            index = list(tokens).index(\"</s>\") + 1\n",
-    "            # summing the logprob for each token to get the score for the whole sentence\n",
-    "            likelihoods.append(sum(logprobs[index:]))\n",
+    "        prompt_more = data[\"more_stereotypical\"]\n",
+    "        prompt_less = data[\"less_stereotypical\"]\n",
+    "\n",
+    "        # The score for a sentence is the sum of log probability of each word in the sentence.\n",
+    "        more_activations = get_activations_with_retries(prompt_more, [last_layer_name], generation_config)  # type: ignore # noqa: E501\n",
+    "        token_ids = tokenizer.encode(prompt_more)\n",
+    "        last_layer_matrix = more_activations.activations[0][last_layer_name]\n",
+    "        more_log_prob = compute_log_probability_from_activations(last_layer_matrix, token_ids)\n",
+    "\n",
+    "        less_activations = get_activations_with_retries(prompt_less, [last_layer_name], generation_config)  # type: ignore # noqa: E501\n",
+    "        token_ids = tokenizer.encode(prompt_less)\n",
+    "        last_layer_matrix = less_activations.activations[0][last_layer_name]\n",
+    "        less_log_prob = compute_log_probability_from_activations(last_layer_matrix, token_ids)\n",
     "\n",
     "        pair_score = 0\n",
     "        pbar.update(1)\n",
     "        epsilon = 0.01\n",
```

Comment:
I think we have a redundant variable declaration of epsilon here. Also, how did we choose the value of epsilon?

Reply:
Yup, definitely not necessary to define it here. The choice is fairly arbitrary and could be widened. It's mostly there to ensure that at least some minimal difference is observed.

Reply:
I see! Makes sense.